core/FrontController.php

```diff
@@ -595,4 +573,49 @@ private function throwIfPiwikVersionIsOlderThanDBSchema()
             throw new DatabaseSchemaIsNewerThanCodebaseException(implode(" ", $messages));
         }
     }
+
+    private function makeAuthenticator()
+    {
+        // if the session cookie exists in this request, validate the session instead of trying to
+        // authenticate the request.
+        $sessionAuthCookieFactory = StaticContainer::get(SessionAuthCookieFactory::class);
+        if ($sessionAuthCookieFactory->isCookieInRequest()) {
+            Session::start();
```

Comment:
This here will start a session for API requests but there should never be any session for API requests see eg https://github.com/piwik/piwik/blob/3.2.0/core/FrontController.php#L404-L409 . There is also a check to start the session only if Piwik is installed which may be good here as well.

Also is it actually needed to start the session twice (here and in dispatch see link above?)

Also: Ideally we close the session directly again if needed as it increases performance because PHP can then execute more requests for the same user in parallel and doesn't have to wait for the other request to finish.

Reply:
I think it might be useful to allow session auth for API methods. We could, for example, remove the token auth from the JS code. (Note: we could also use something like a JWT, but since there’s no server side component, they’d be less secure than a session. Ie, a stolen JWT would be usable by an attacker.)

What about not starting the session, if there’s a token auth? So API requests will start a session ONLY if the session cookie is present.

Reply:
Sessions should not be used at all when using the API. Also not for authentication. I know we are not doing REST but being stateless is like one of the core concepts there and should be done here as well. It is like a best practice for APIs.

Reply:
You’re right, I was optimizing for the wrong metric.